dev/core#1280 Fix ContributionPage soft_credit translation

mlutfy opened the pull request:
Overview
----------------------------------------

https://lab.civicrm.org/dev/core/issues/1280

To reproduce:

* Enable multilingual (Admin > Localisation > languages, checkbox at the bottom)
* Add a second language (Admin > Localisation > Languages, 'available languages')
* Create a Contribution Page
  * Enable "Honoree Section Enabled"
  * Enter a title/description, types
  * save

Now switch languages, edit text and save.

Result: data disappears or in wrong language.

![inhonor-of-translation](https://user-images.githubusercontent.com/254741/76980956-45395580-6910-11ea-8556-4c9daa0bc8a9.gif)

mlutfy commented:
jenkins, test this please

demeritcowboy commented on tests/phpunit/CRM/Contribute/Form/ContributionPageTranslationTest.php:
```diff
@@ -0,0 +1,91 @@
+    $this->assertEquals($json['soft_credit']['en_US']['honor_block_title'], 'In Honor Title EN');
+    $this->assertEquals($json['soft_credit']['en_US']['honor_block_text'], 'In Honor Text EN');
+    $this->assertEquals($json['soft_credit']['fr_FR']['honor_block_title'], 'In Honor Title FR');
+    $this->assertEquals($json['soft_credit']['fr_FR']['honor_block_text'], 'In Honor Text FR');
```
I like the test - just these params are backwards. Expected is the first param. When it fails it's briefly confusing.

mlutfy replied:
fixed

demeritcowboy commented on CRM/Contribute/Form/ContributionPage/Settings.php:
```diff
@@ -348,8 +348,16 @@ public function postProcess() {
+        $ufJoinDAO = new CRM_Core_DAO_UFJoin();
+        $ufJoinDAO->module = $ufJoinParam['module'];
+        $ufJoinDAO->entity_id = $ufJoinParam['entity_id'];
+        $ufJoinDAO->find(TRUE);
```
Is it possible that soft-credit profiles could be used for something other than contribution pages? If so, you'd want to take entity_table into account. It used to before with the deleteAll. Otherwise this might match on multiple and might update the wrong one.

mlutfy replied:
I think the combo would be surprising, but I added an extra check, just in case.
However, I only did a quick code edit, need to re-test.

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
@demeritcowboy what's your take on this - it looks like you reviewed it - is it pending a response from @mlutfy 
?

eileenmcnaughton commented:
@mlutfy style issues...

demeritcowboy commented:
I don't remember running it just code review. Yes, one clerical, one functional.

mlutfy commented:
Keeping a copy in case it disappears too quickly:

![Capture d’écran de 2020-07-16 18-27-08](https://user-images.githubusercontent.com/254741/87728854-facac700-c791-11ea-8a7a-e038a87a442f.png)

mlutfy commented:
Thanks for the review @demeritcowboy !

I updated my PR with the recommandations:

- switched the order of the test params
- added an extra check for the ufParams (entity_table)
- fixed the styling

I just need to do one last test to make sure that adding the extra check didn't break anything.

mlutfy commented:
I did a bit more testing and it seems to be working as expected.

* http://core-16838-552ok.test-3.civicrm.org:8001/
* admin / aZ7Ky4RO8qNN

eileenmcnaughton commented:
@mlutfy please self-merge once @demeritcowboy gives this a thumbs up

demeritcowboy commented:
@mlutfy I think you've found an awesomely sneaky way of getting tests to pass. Just name the file slightly differently than the actual test class. The test framework won't even run it! Easy pass. (FYI @totten @seamuslee001 )

So either the file or the classname needs to be renamed, but otherwise all good.

mlutfy commented:
Yikes, good catch :grimacing: 

Note to self: in the full test log, ctrl-F for ContributionPageTranslationTest

demeritcowboy commented:
Looks good. So I think this is now "merge-on-pass".

mlutfy commented:
Thanks again, @demeritcowboy, very much appreciated!

mlutfy commented:
Tests have not completed yet, but seeing this now:

```
ok 600 - CRM_Contribute_Form_ContributionPageTranslationTest::testCreateHonor
```

demeritcowboy commented:
Cool.

And actually if you have a minute https://github.com/civicrm/civicrm-core/pull/17806 just needs an r-run and that's the last piece before the outputhandler stuff is ready.
